Emit telemetry events on success and failure of the Create Job, Create Job Definition, Create Job from Job Definition hooks

andrii-i opened the pull request:
Emit telemetry events on success and failure of the Create Job, Create Job Definition, Create Job from Job Definition hooks. 
* Log object's `body.name` string is now following a new convention: `org.jupyter.jupyter-scheduler.SOURCE.EVENT[.STATUS]` (vs old convention `org.jupyter.jupyter-scheduler.SOURCE.EVENT`) where `STATUS` is either `"success"` or `"failure"`
* `CreateJob` now emits `"create-job"` or `"create-job-definition"` for `EVENT` instead of just `"create"` depending on if job or job definition is being created
* Log object's `body` now has optional `detail` parameter storing error message (or other string with event detail)
```ts
type EventLog = {
  body: { name: string; detail?: string };
  timestamp: Date;
};
```
* Improve error catching logic by catching errors as `unknown` type and using type guarding instead of assuming all errors caught would be of `Error` type (ts runtime error type) which is not guaranteed.
* Fix #471 

Emitted objects examples:
<img width="480" alt="Screenshot 2024-01-18 at 2 52 01 PM" src="https://github.com/jupyter-server/jupyter-scheduler/assets/26686070/e8b40131-e16a-4afa-8a5d-309adcbb7740">

<img width="481" alt="Screenshot 2024-01-18 at 2 50 31 PM" src="https://github.com/jupyter-server/jupyter-scheduler/assets/26686070/823ff018-7dc8-4a94-ba07-f4990b043ba3">

bhadrip commented on src/mainviews/create-job-from-definition.tsx:
```diff
@@ -139,6 +139,7 @@ export function CreateJobFromDefinition(
         createJobFromDefinitionModel
       )
       .then(response => {
+        log('create-job-from-definition.create-job.success');
```
can we log the error response - http response code and message ?

andrii-i replied:
Added optional `error` parameter with error message and HTTP status code to log object's `body` : https://github.com/jupyter-server/jupyter-scheduler/pull/472/files#diff-0b5adbfe7b36e4ae2f479291e20152e33e940f7f265162d77f40f6bdb5da7405R44-R47

3coins commented on src/index.tsx:
```diff
@@ -41,6 +41,11 @@ export namespace CommandIDs {
 export { Scheduler } from './tokens';
 
+type EventLog = {
+  body: { name: string; error?: { message?: string; httpStatusCode?: number } };
```
Capturing the error detail/string should be sufficient to capture the error.
```suggestion
  body: { name: string; detail?: string } };
```

3coins commented on src/context.ts:
```diff
@@ -1,10 +1,12 @@
-});
+export type Logger = (eventName: string, error?: Error) => void;
+export const LogContext = React.createContext<Logger>(
+  (eventName: string, error?: Error) => {
```
```suggestion
  (eventName: string, eventDetail?: string) => {
```

3coins commented on src/context.ts:
```diff
@@ -1,10 +1,12 @@
-export const LogContext = React.createContext<Logger>((eventName: string) => {
-  /*noop*/
-});
+export type Logger = (eventName: string, error?: Error) => void;
```
Making error as `Error` type is too specific to HTTP errors. Though the intent here is to primarily capture the HTTP errors, having something simpler can accomplish this. 

```suggestion
export type Logger = (eventName: string, eventDetail?: string) => void;
```

andrii-i commented:
@3coins thank you for looking into this, implemented your comments re not logging an `httpResponseCode`. Also added improved error catching logic as discussed.

JasonWeill reviewed (commented):
"Unknown error occurred" is a generic message, but it may not inspire confidence. Do we have any idea about what sorts of errors might trigger the generic message? If so, can we provide any additional guidance? Something like, "An error occurred. Please try again."?

andrii-i commented:
@JasonWeill thank you for looking into this. Any error / error object that does not have fields expected from [es5 Error object](https://github.com/microsoft/TypeScript/blob/1982349339b9c1fd78c93195d572ffc6fd5041b0/src/lib/es5.d.ts#L1057) would trigger the generic error message. Most probably to occur are connection failures, connection timeouts, third-party libraries throwing something else than es5 `Error` (other objects or just strings). So the message needs to be generic, I think the message that you proposed works well, updated the function to use it.

JasonWeill reviewed (approved):
Looks good, thanks!
